include/boost/geometry/algorithms/detail/is_valid/pointlike.hpp

```diff
@@ -42,14 +45,26 @@ struct is_valid<Point, point_tag>
 // (have identical coordinate values in X and Y)
 //
 // Reference: OGC 06-103r4 (6.1.5)
-template <typename MultiPoint>
-struct is_valid<MultiPoint, multi_point_tag>
+template <typename MultiPoint, bool AllowEmptyMultiGeometries>
+struct is_valid<MultiPoint, multi_point_tag, AllowEmptyMultiGeometries>
 {
-    static inline bool apply(MultiPoint const& /*multipoint*/)
+    template <typename VisitPolicy>
+    static inline bool apply(MultiPoint const& multipoint,
+                             VisitPolicy& visitor)
     {
-        // we allow empty multi-geometries, so an empty multipoint
-        // is considered valid
-        return true;
+        if (AllowEmptyMultiGeometries || boost::size(multipoint) > 0)
+        {
+            // we allow empty multi-geometries, so an empty multipoint
+            // is considered valid
+            visitor.template apply<no_failure>();
+            return true;
+        }
+        else
+        {
+            // we do not allow an empty multipoint
+            visitor.template apply<failure_few_points>();
+            return false;
```

Comment:
I don't know if it is a good idea, but just brainstorm: we have this pattern often, visitor.template ... no_failure, return true, or visitor.template failure, return false. Can we not just return visitor.template apply<....> ? So then the apply with no-failure always returns true and the rest false? Or does this have drawbacks?
Anyway, if good idea, you don't have to apply that now, we can postpone that action.